### DISCUSSION TITLE
How does PostLongForm work for questions and comments?

### DISCUSSION THREAD

**Discussion by taytayp** [ialbert/biostar-central#915] [Q&A]
This is less an issue more a feature I am trying to implement, so a discussion topic feels more appropriate.

I am running my own instance of the biostars forum for use within our school, and we are interested in having an "anonymous" post feature. A user can select a checkbox on submission/edit and their account will be hidden from everyone besides moderators when posting.

I have managed to implement it successfully within answers and original posts, adding a field to PostLongForm and PostShortForm inside `biostar/forum/forms.py` and adding a column to the database to record this boolean result. 

What I can't seem to figure out is how comments _on_ posts and answers work, since I want this to be anonymous too. They seem to also use the PostLongForm class but I can't seem to figure out how they differ from full posts/questions and it is driving me slowly insane.

I have made a number of changes so forgive the lack of examples here, but if you want to get into the weeds with me please check out my recent commits [here](https://gitlab.com/taylor.f/biostar-central).

Any guidance would be appreciated!

**Comment by Natay** [comment]
Hey @taytayp took a peek your commits and you have all the right pieces -- here are some pointers to help clarify.

Main difference is comments use `ajax` to handle data asynchronously and in place. This is why clicking 'Add Comment' only loads that portion of the page, while original answers/questions use a regular request-response cycle that refreshes the whole thing. 


pointers :
- > They seem to also use the PostLongForm class but I can't seem to figure out how they differ from full posts/questions and it is driving me slowly insane.
 
   Small correction,  comments use `PostShortForm` for creating in [`biostar.forum.ajax.ajax_comment_create`](https://github.com/Natay/biostar-central/blob/master/biostar/forum/ajax.py#L294-L314) and for editing in [`biostar.forum.ajax.ajax_edit`](https://github.com/Natay/biostar-central/blob/master/biostar/forum/ajax.py#L243-L272) .

- The field you want to set ( `post.anon` )  needs to be passed to the ajax functions that call `biostar.forum.ajax.ajax_comment_create` ( [create_comment](https://github.com/Natay/biostar-central/blob/master/biostar/forum/static/inplace.js#L18))  and `biostar.forum.ajax.ajax_edit` ( [edit_comment](https://github.com/Natay/biostar-central/blob/master/biostar/forum/static/inplace.js#L192) ) . I would look at other fields to see how they are set and mirror them ( essentially need to parse the form fields [here](https://github.com/Natay/biostar-central/blob/master/biostar/forum/static/inplace.js#L20) and pass it to the `data` being sent [here](https://github.com/Natay/biostar-central/blob/master/biostar/forum/static/inplace.js#L35)). Needless to say, you will need to include `post.anon` in the [inplace form ](https://github.com/Natay/biostar-central/blob/master/biostar/forum/templates/forms/form_inplace.html)

[Quick guide on using `ajax` with django](https://www.pluralsight.com/guides/work-with-ajax-django)

**Comment by taytayp** [comment]
Thanks for the tips @Natay ! I think I have it mostly working minus one exception. I am unable to get the checkbox's value within the `inplace.js` script. I try to find it with `$('#id-anon').is(':checked');` but it looks to be blank (even if I `console.log` just the `$('#id-anon') object). Any suggestions here? My JS knowledge is definitely on the lower side compared to my python haha.

Pushed changes to the GitLab repo above if you want to see my changes. Thanks again!

**Comment by tapanrathod** [comment]
In the Biostar codebase, comments actually use PostShortForm (or CommentForm), whereas top-level posts and answers rely on PostLongForm. Both forms inherit from a base post form class, but comments intentionally strip out long-form attributes like title fields and tags.

​When processing form submissions in biostar/forum/views.py, check the view handling comment creation (typically post_create or comment_create). Because comments render through a slightly different form instance, you need to explicitly bind your new anonymous boolean field to both PostLongForm and PostShortForm inside forms.py. Once added to both form classes, pass that field value to the underlying Post model creation call during comment submission.

**Comment by ialbert** [comment]
A new version of Biostar is being developed. See:

https://github.com/ialbert/biostar-central/discussions/961

and:

https://test.biostars.org/
